local dictionary = require("../../../utilities/dictionary")

export type Mapping = {
	text: string,
	mapping: { string },
}

local Elements: { [string]: Mapping } = {
	H = { text = "22", mapping = { "H" } },
	He = { text = "13112221133211322112211213322112", mapping = { "Hf", "Pa", "H", "Ca", "Li" } },
	Li = { text = "312211322212221121123222112", mapping = { "He" } },
	Be = { text = "111312211312113221133211322112211213322112", mapping = { "Ge", "Ca", "Li" } },
	B = { text = "1321132122211322212221121123222112", mapping = { "Be" } },
	C = { text = "3113112211322112211213322112", mapping = { "B" } },
	N = { text = "111312212221121123222112", mapping = { "C" } },
	O = { text = "132112211213322112", mapping = { "N" } },
	F = { text = "31121123222112", mapping = { "O" } },
	Ne = { text = "111213322112", mapping = { "F" } },
	Na = { text = "123222112", mapping = { "Ne" } },
	Mg = { text = "3113322112", mapping = { "Pm", "Na" } },
	Al = { text = "1113222112", mapping = { "Mg" } },
	Si = { text = "1322112", mapping = { "Al" } },
	P = { text = "311311222112", mapping = { "Ho", "Si" } },
	S = { text = "1113122112", mapping = { "P" } },
	Cl = { text = "132112", mapping = { "S" } },
	Ar = { text = "3112", mapping = { "Cl" } },
	K = { text = "1112", mapping = { "Ar" } },
	Ca = { text = "12", mapping = { "K" } },
	Sc = { text = "3113112221133112", mapping = { "Ho", "Pa", "H", "Ca", "Co" } },
	Ti = { text = "11131221131112", mapping = { "Sc" } },
	V = { text = "13211312", mapping = { "Ti" } },
	Cr = { text = "31132", mapping = { "V" } },
	Mn = { text = "111311222112", mapping = { "Cr", "Si" } },
	Fe = { text = "13122112", mapping = { "Mn" } },
	Co = { text = "32112", mapping = { "Fe" } },
	Ni = { text = "11133112", mapping = { "Zn", "Co" } },
	Cu = { text = "131112", mapping = { "Ni" } },
	Zn = { text = "312", mapping = { "Cu" } },
	Ga = { text = "13221133122211332", mapping = { "Eu", "Ca", "Ac", "H", "Ca", "Zn" } },
	Ge = { text = "31131122211311122113222", mapping = { "Ho", "Ga" } },
	As = { text = "11131221131211322113322112", mapping = { "Ge", "Na" } },
	Se = { text = "13211321222113222112", mapping = { "As" } },
	Br = { text = "3113112211322112", mapping = { "Se" } },
	Kr = { text = "11131221222112", mapping = { "Br" } },
	Rb = { text = "1321122112", mapping = { "Kr" } },
	Sr = { text = "3112112", mapping = { "Rb" } },
	Y = { text = "1112133", mapping = { "Sr", "U" } },
	Zr = { text = "12322211331222113112211", mapping = { "Y", "H", "Ca", "Tc" } },
	Nb = { text = "1113122113322113111221131221", mapping = { "Er", "Zr" } },
	Mo = { text = "13211322211312113211", mapping = { "Nb" } },
	Tc = { text = "311322113212221", mapping = { "Mo" } },
	Ru = { text = "132211331222113112211", mapping = { "Eu", "Ca", "Tc" } },
	Rh = { text = "311311222113111221131221", mapping = { "Ho", "Ru" } },
	Pd = { text = "111312211312113211", mapping = { "Rh" } },
	Ag = { text = "132113212221", mapping = { "Pd" } },
	Cd = { text = "3113112211", mapping = { "Ag" } },
	In = { text = "11131221", mapping = { "Cd" } },
	Sn = { text = "13211", mapping = { "In" } },
	Sb = { text = "3112221", mapping = { "Pm", "Sn" } },
	Te = { text = "1322113312211", mapping = { "Eu", "Ca", "Sb" } },
	I = { text = "311311222113111221", mapping = { "Ho", "Te" } },
	Xe = { text = "11131221131211", mapping = { "I" } },
	Cs = { text = "13211321", mapping = { "Xe" } },
	Ba = { text = "311311", mapping = { "Cs" } },
	La = { text = "11131", mapping = { "Ba" } },
	Ce = { text = "1321133112", mapping = { "La", "H", "Ca", "Co" } },
	Pr = { text = "31131112", mapping = { "Ce" } },
	Nd = { text = "111312", mapping = { "Pr" } },
	Pm = { text = "132", mapping = { "Nd" } },
	Sm = { text = "311332", mapping = { "Pm", "Ca", "Zn" } },
	Eu = { text = "1113222", mapping = { "Sm" } },
	Gd = { text = "13221133112", mapping = { "Eu", "Ca", "Co" } },
	Tb = { text = "3113112221131112", mapping = { "Ho", "Gd" } },
	Dy = { text = "111312211312", mapping = { "Tb" } },
	Ho = { text = "1321132", mapping = { "Dy" } },
	Er = { text = "311311222", mapping = { "Ho", "Pm" } },
	Tm = { text = "11131221133112", mapping = { "Er", "Ca", "Co" } },
	Yb = { text = "1321131112", mapping = { "Tm" } },
	Lu = { text = "311312", mapping = { "Yb" } },
	Hf = { text = "11132", mapping = { "Lu" } },
	Ta = { text = "13112221133211322112211213322113", mapping = { "Hf", "Pa", "H", "Ca", "W" } },
	W = { text = "312211322212221121123222113", mapping = { "Ta" } },
	Re = { text = "111312211312113221133211322112211213322113", mapping = { "Ge", "Ca", "W" } },
	Os = { text = "1321132122211322212221121123222113", mapping = { "Re" } },
	Ir = { text = "3113112211322112211213322113", mapping = { "Os" } },
	Pt = { text = "111312212221121123222113", mapping = { "Ir" } },
	Au = { text = "132112211213322113", mapping = { "Pt" } },
	Hg = { text = "31121123222113", mapping = { "Au" } },
	Tl = { text = "111213322113", mapping = { "Hg" } },
	Pb = { text = "123222113", mapping = { "Tl" } },
	Bi = { text = "3113322113", mapping = { "Pm", "Pb" } },
	Po = { text = "1113222113", mapping = { "Bi" } },
	At = { text = "1322113", mapping = { "Po" } },
	Rn = { text = "311311222113", mapping = { "Ho", "At" } },
	Fr = { text = "1113122113", mapping = { "Rn" } },
	Ra = { text = "132113", mapping = { "Fr" } },
	Ac = { text = "3113", mapping = { "Ra" } },
	Th = { text = "1113", mapping = { "Ac" } },
	Pa = { text = "13", mapping = { "Th" } },
	U = { text = "3", mapping = { "Pa" } },
}

return {
	Elements = Elements,
	ByText = dictionary.map(Elements, function(element, elementName)
		return elementName, element.text
	end),
}
